### PR TITLE
A lot of new actions and functionality for the NAPALM pack

### PR DESCRIPTION
**Pull request by robwdwd**
I've taken the initial pack by mierdin and added actions for most napalm functions and some action chains/workflows to act on configuration changes on devices, bgp changes and interface events.

Updated the documentation and provides some examples on how to link logstash, rsyslog together to get events from the network devices to pass to stackstorm.

### CODE REVIEW

**Comment by LindsayHill** [comment]
This looks like some great work @robwwd - thank you!

File: actions/chains/bgp_prefix_exceeded_chain.yaml
Comment (LindsayHill) [inline actions/chains/bgp_prefix_exceeded_chain.yaml]: So I can see we're going to get into an argument here with our American English-speaking cousins. How ready are you to defend neighbour vs neighbor? :trollface: 

(For reference: I grew up speaking New Zealand English, which is much closer to the Queen's English than this derivative I now battle on a daily basis)

Comment (robwdwd) [reply actions/chains/bgp_prefix_exceeded_chain.yaml]: Her majesty would have my head if I spelt it any other way (except where I've cut and pasted in the other actions ;) ). I too have had this battle most of my career and am thankful that I can abbreviate it on most occasions as 'nei' and not have to wear out my backspace key to delete the u.

Comment (tonybaloney) [reply actions/chains/bgp_prefix_exceeded_chain.yaml]: @robwwd seconded. Prince Charles himself was one of the original authors of NAPALM and he would be most upset if American English were used.

**Comment by Mierdin** [comment]
@robwwd Hey Rob - just wanted to check in and tell you two things. First, thanks so much for this PR! Lots of awesome work here, and I can tell you this is a big priority for me, so I'll be working with you to get this merged ASAP.

Second, and this is sort of related to the first point - we're in the middle of releasing 2.2 of StackStorm, and I am in charge of this release, so my priorities have shifted during that period of time. We're hoping to have this finished this week, so hopefully you'll hear something from me before Friday, or maybe early next week.

Thanks again!

File: actions/chains/bgp_prefix_exceeded_chain.yaml
Comment (Mierdin) [reply actions/chains/bgp_prefix_exceeded_chain.yaml]: So.......not trying to start a war here, but in this case I would prefer that we stick to the naming conventions used in the NAPALM project itself. Throughout the project, the word `neighbor` is used.

Please direct all complaints to @dbarrosop - I'm just the messenger :)

File: README.md
Comment (Mierdin) [inline README.md:222]: Let's put this right at the top, with a highlighted section. This is a lot of new functionality and I want folks to know that we're actively interested in putting this pack through it's paces. Something like "This pack is actively being developed and should be considered BETA status for the time being. Please open a github issue or pull request if you run into any problems with this pack"

File: actions/chains/bgp_prefix_exceeded_chain.yaml
Comment (robwdwd) [reply actions/chains/bgp_prefix_exceeded_chain.yaml]: Changed it now ;)

Comment (Mierdin) [reply actions/chains/bgp_prefix_exceeded_chain.yaml]: Thanks, though I'm thinking this needs to be changed in the entire PR. Again, I personally don't really care one way or another in terms of spelling, but what I do care about is consistency - and the existing NAPALM users will expect the spelling without the u.

File: actions/get_arp_table.yaml
Comment (Mierdin) [inline actions/get_arp_table.yaml]: Should be "ARP"

File: actions/get_log.py
Comment (Mierdin) [inline actions/get_log.py]: This works - generally I prefer to store this kind of thing in a dict and do a lookup. Like this:

```
{
    'iosxr': {
        'log_cmd': 'show log',
        'commands': ['term width 0', 'term len 0']
    }
}
```

**Review by Mierdin — CHANGES_REQUESTED** [review]
Rob - this is a phenomenal PR. Thanks so much for working on this. I want to get this merged ASAP so I only have a few general points, and a handful of inline comments. Here are some general thoughts:

- As I mention in an inline comment, I don't personally care too much about "neighbor" vs "neighbour", but I do care about consistency, and the NAPALM project has set a precedent that the spelling "neighbor" is used. So I would prefer that all of the references in this PR are changed to that
- I am curious about the mix of actionchains and mistral workflows. Is there a particular reason you used both? This isn't a dealbreaker, but if I were to choose, I'd go 100% mistral - wondering why you mixed.
- See my inline comment - I'd like to see a section near the very top of the README that mentions this is in active development, and should be considered BETA for the time being - mention that PRs and issues are encouraged. This has nothing to do with the quality of your submission and more about attracting more people to get eyes on it.
- There is quite a bit of duplicate code in the python actions. I would like to see some of the housekeeping stuff like looking up creds/devices and getting a connection to devices inside the base class so that it can be re-used.
- Also, and this is mostly for me - I’m curious what you tested this with - I want to run this on a few devices of my own and would like to pick devices you didn’t
- I am not sure if we're able to use icon.png - I have messaged the Napalm contributors to ask about this.

Again, great work. I want to get this merged ASAP so that we can get this into the hands of the community and get their feedback. Let us know if there's anything we can do to help.

File: README.md
Comment (robwdwd) [reply README.md:222]: Good idea, I've updated this now with that at the top.

File: actions/get_log.py
Comment (robwdwd) [reply actions/get_log.py]: Agreed, that seems a much better way of doing things.

**Comment by robwdwd** [comment]
- Thanks for all the positive feedback and for taking the time to review things, apprecited and always good to get such good and detailed feedback.
- No worries on the spelling of neighbours, I did try and keep the the NAPALM spelling but some got through there I used neighbor in the file naming but then in the code I used neighbour in places
- I'd done action chains for the BGP one because originally I was using some looping with Jinja (in the email output) which I couldn't replicate with mistral. But then I've changed  it and used (although I'm not entirely happy with how I did it that way either) the action to grab HTML headers and then in the code generated HTML output. I used mistral for the other stuff because I could generate a fail where as in the action chain if I send a failure email the action change succeeds (because the email action is a success) and I could not find a way to force a fail. I found for outputing some kind of email or message somewhere Jinja gave me more flexibility 
- I've tested it on IOS, IOS-XR, JunOS and EOS boxes. Of various sizes. JunOS mostly does a good job but the IOS-XR can be a bit flakey in some situations (for example when a device has thousands of interfaces on it it tends to timeout - but not always).

**Comment by Mierdin** [comment]
Okay I took the liberty of doing some of the heavy lifting re: centralizing that duplicate code in my branch [here](https://github.com/StackStorm-Exchange/stackstorm-napalm/compare/centralize-setup). Give that a good sanity check then I'd love it if you merged that into your `further-actions-pull` branch so that it's part of this PR.

Whats left is mostly cosmetic. I noticed a few errors in some descriptions that need fixed but nothing earth-shattering. I updated the README to be a bit clearer on usage, so the user knows what needs to be in the config and how to override these in the command-line at runtime. Feel free to take that and run with it.

Also, my big TODO for tomorrow is in depth testing. In addition to the fact that I haven't done it yet (thanks for your first round of this, sounds like you tested against a lot of devices) a new round of detailed testing is warranted due to my changes.

Let me know if you have questions/concerns.

**Comment by dbarrosop** [comment]
Hi, on behalf of the napalm-automation community. This is great :)

Feel free to reach to us (happy to be the liason or just open an issue in our main "napalm" repo) if there is something you want us to look into or if you think there might be something interesting we could add/change to integrate better with st2.

**Comment by robwdwd** [comment]
Looks great and thanks for taking the time to centralise some of the code there.

Although I see you removed the host_ip parameter. I did the host IP as not all hosts resolve right from the syslog logsource and host.

It was covering broken DNS and broken syslog messages (mainly from cisco boxes). So a syslog would come in from a host identifying itself as peering1.LON with IP 10.1.1.1, reverse DNS for 10.1.1.1 had not been updated but points to core4.NYC. It's a Cisco box so the syslog server (in my case rsyslog) is doing a reverse lookup to get the host because it's not specifically telling syslog in the syslog messages it's peering1.LON. So now stackstorm is triggering an event against core4.NYC. Giving stackstorm an IP address to connect to allows it to ignore the fact that it's not really core4.NYC.

So in essence the IP address was what we always want to work with (it should always be correct in the syslog message) and the hostname is just there for the driver lookup and identification purposes (although if the wrong host is a different OS then it's an issue but can be fixed in DNS). This way, for example, you avoid an interface down event triggering an actionchain to reset an interface on the wrong device. 

The lookup of the hostname was there when an action was run manually to save whomever was running it from doing the lookup themselves. But for syslog triggered events (in my case I was using rsyslog->logstash->st2 web hook.) you have to trust the source IP of the syslog message.

**Comment by Mierdin** [comment]
The first reason I removed host_ip was that it was pretty confusing to have both `host_ip` and `hostname` as parameters to these actions. Again, I want to make sure this pack is consistent with the underlying library whenever possible, and `host_ip` will look strange to anyone that's used NAPALM before.

But besides that, I'm not convinced that solving what effectively boils down to a DNS problem in this pack is the right solution. I'd rather these actions just accept a hostname parameter, and assume it's correct.

**Comment by robwdwd** [comment]
Don't agree on that one. No need to rigidly stick with napalm library parameters as this is stackstorm. 

Host ip is optional but gives more flexibility. There may be a number of reasons why DNS doesn't resolve but you still want a device name. You could do everything with ip in the hostname field and have a separate identifier. 

If you have 50 thousands CPE devices you trust the source Ip not the hostname. 

Another good example is JunOS where the Device reports a configured hostname plus the routing engine. Core1-re0 but not a fqdn so what do you do with reverse dns then? What if it fails over to re1? So you cant trust the hostname here so you'd want the IP source. But you still want a hostname for human eyes.

**Comment by Mierdin** [comment]
So to be clear, I agree that these are problems - my point is only that these problems should be solved outside this pack. This approach worked for you, but what about others that may want it solved in a different way?

Also, the `hostname` param of NAPALM's `get_network_driver` function doesn't need to be a hostname or FQDN - it could also be an IP address. There's nothing preventing the user to populate the configuration with IP addresses for the `hostname` param. In my mind, this obviates the need for a separate `host_ip` parameter.

I could be totally wrong, but that's my $0.02 - I'll let others chime in here and see if we can get some consensus.

**Comment by Mierdin** [comment]
@dbarrosop Thanks for stopping by! We will reach out if we need anything - please do not hesitate to leave comments on any work on this pack - would appreciate your insights.

**Comment by bigmstone** [comment]
I tend to agree with @Mierdin here. `hostname` is universally accepted in most libraries to mean `ip`, `short-name`, or `fqdn`. If we want to just name it `host` that might be a good middle ground, but I'm -1 on two params for the same value unless there's some good technical reason to do it.

**Comment by lakshmi-kannan** [comment]
I am in favor of `host` as the parameter name as opposed to `hostname` which can be confusing. But I am definitely -1 to having two parameters.

**Comment by robwdwd** [comment]
My opinion is that the ip is optional to overwrite DNS if needed.  I know the hostname can also be an IP. However...

Given this is automation we are talking about we need to consider how the inputs are presented. 

Taking syslog as an example you are going to get a logsource and hostname. The issue is hostname can be set by syslog or by the device depending on the syslog format the device is sending in. Cisco boxes don't set it so syslog does a reverse lookup on the source ip. Juniper boxes do send it but it has the RE tag. 

So in that situation it makes more sense to go with the logsource (assuming syslog is configured to pass the initiating device ip to the final syslog host feeding stackstorm). This works better if it's an IP but can be a hostname if syslog is looking things up. 

If we rely on ip addressing when someone looks at an event they need to correlate that ip manually with a device.

So if we look at the middle ground. Rather than make and optional ip parameter we keep hostname but add a optional host which can be used to identify a device in a way which a human can correlate. Providers usually have a naming scheme to make support staff jobs easier. 

I think we need to be flexible in dealing with as many use cases as possible. Most use cases would take input from syslog and SNMP.

File: actions/cli.yaml
Comment (lakshmi-kannan) [inline actions/cli.yaml:12]: Position argument is used only for scripts that that take positional arguments. In this case, cli.py seems like a python action which takes in named parameters.

**Comment by bigmstone** [comment]
@robwwd If you're trying to save a step for looking it up then why not feed it back in action output? `ip = device_connected.ip`

This seems more pure than two options. It's unclear to the user which wins if they don't match and not very reliable if, in fact, they don't match. Better to have the script send this as output back to the user where it can be relied on. Same with hostname if you want to attempt to look it up and send it back as `hostname` result.

**Comment by bigmstone** [comment]
Also, from an ST2 perspective this is perfect candidate to be wrapped into a workflow. First task can distill which IP/Hostname to use and also log the results, then execute the action using `hostname` or `host` param (whatever we decide to call it). This would give everyone what they want. @robwwd you get the output and inputs you'd desire without the confusion of two very similar inputs.

**Comment by bigmstone** [comment]
Here's an example of what that might look like, just to be explicit:

```
version: '2.0'

robwwd_automation.pre_napalm:
    description: A workflow to run prior to napalm actions to ensure proper device is connected to and logging of hostname.
    type: direct
    input:
        - hostname
        - ip
    tasks:
        get_host:
            action: robwwd_automation.get_host
            input:
                hostname: <% $.hostname %>
                ip: <% $.ip %>
            publish:
                host: <% task(get_host).result.result.host %>
        napalm_action:
            action: napalm.action
                host: <% $.host %>
 ```

**Comment by robwdwd** [comment]
If the consensus is to go with hostname only I'm happy to go with that. 

I can probably work around the issues before st2.

**Comment by Mierdin** [comment]
@robwwd You can still do this inside st2 using the example workflow that @bigmstone posted. With your own action (outside of this NAPALM pack) you can do whatever translations you need, and feed the result into the necessary NAPALM action. This way you get what you need, and the NAPALM pack stays pure.

So you want to merge my branch into yours and then we can merge yours into `master`? I'm thinking there shouldn't be any glaring issues that can't wait until after we've merged this initial pack.

**Comment by Mierdin** [comment]
Thanks for the merges @robwwd - looks like I forgot to update cli.py in my changes. 

I'm inclined to merge this now, since you've already done a lot of work on this, and that way I can quickly fix the issues I caused in a separate PR (unless anyone has objections).

**Pull request merged**
